src/nexgen/nxs_utils/Detector.py

```diff
@@ -0,0 +1,173 @@
+
+from dataclasses import dataclass, field
+from typing import List, Literal, Tuple, Union
+
+from dataclasses_json import dataclass_json
+
+from ..utils import Point3D
+from .Axes import Axis
+
+__all__ = ["EigerDetector", "TristanDetector", "Detector"]
+
+
+class UnknownDetectorTypeError(Exception):
+    pass
+
+
+EIGER_CONST = {
+    "flatfield": "flatfield",
+    "flatfield_applied": "_dectris/flatfield_correction_applied",
+    "pixel_mask": "mask",
+    "pixel_mask_applied": "_dectris/pixel_mask_applied",
+    "bit_depth_readout": "_dectris/bit_depth_readout",
+    "detector_readout_time": "_dectris/detector_readout_time",
+    "threshold_energy": "_dectris/threshold_energy",
+    "software_version": "_dectris/software_version",
+    "serial_number": "_dectris/detector_number",
+}
+
+TRISTAN_CONST = {
+    "flatfield": "Tristan10M_flat_field_coeff_with_Mo_17.479keV.h5",
+    "flatfield_applied": False,
+    "pixel_mask": "Tristan10M_mask_with_spec.h5",
+    "pixel_mask_applied": False,
+    "software_version": "1.1.3",
+    "detector_tick": "1562.5ps",
+    "detector_frequency": "6.4e+08Hz",
+    "timeslice_rollover": 18,
+}
+
+
+@dataclass_json
+@dataclass
+class EigerDetector:
+    description: str
+    image_size: List[float] | Tuple[float]
+    sensor_material: Literal["Si", "CdTe"]
+    overload: int
+    underload: int
+    pixel_size: List[str | float] = field(
+        default_factory=lambda: ["0.075mm", "0.075mm"]
+    )
+    detector_type: str = "Pixel"
+
+    @property
+    def sensor_thickness(self) -> str:
+        if self.sensor_material == "Si":
+            return "0.450mm"
+        else:
+            return "0.750mm"
+
+
+@dataclass_json
+@dataclass
+class TristanDetector:
+    description: str
+    image_size: List[float] | Tuple[float]
+    sensor_material: str = "Si"
+    sensor_thickness: str = "0.5mm"
+    pixel_size: List[str | float] = field(
+        default_factory=lambda: ["5.5e-05m", "5.5e-05m"]
+    )
+    detector_type: str = "Pixel"
+    mode: Literal["events", "images"] = "events"
+
+
+@dataclass_json
+@dataclass
+class SinglaDetector:
+    description: str
+    image_size: List[float] | Tuple[float]
+    sensor_material: str = "Si"
+    sensor_thickness: str = "0.450mm"
+    overload: int = 199996
+    underload: int = -1
+    pixel_size: List[str | float] = field(
+        default_factory=lambda: ["0.075mm", "0.075mm"]
+    )
+    detector_type: str = "HPC"
+
+
+DetectorType = Union[EigerDetector, TristanDetector, SinglaDetector]
+
+
+class Detector:
+    def __init__(
+        self,
+        detector_params: DetectorType,
+        detector_axes: List[Axis],
+        beam_center: List[float],
+        exposure_time: float,
+        module_vectors: List[Point3D] | List[Tuple],
+    ):
+        if type(detector_params) not in [
+            EigerDetector,
+            TristanDetector,
+            SinglaDetector,
+        ]:
+            raise UnknownDetectorTypeError("Unknown detector.")
```

Comment:
This should be covered by the type system, I don't think we need to check it twice